File: 0601-0700/0687-Longest Univalue Path/0687-Longest Univalue Path.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int longestUnivaluePath(TreeNode* root) {
        int maxLen = 0;
        longestPath(root, maxLen);
        return maxLen;
    }
    
private:
    int longestPath(TreeNode* root, int& maxLen) {
        if (root == nullptr) {
            return -1;
        }
        
        int leftLen = longestPath(root->left, maxLen);
        int rightLen = longestPath(root->right, maxLen);
        int len = 0;
        int currLen = 0;
        if (root->left && root->left->val == root->val) {
            len = leftLen + 1;
            currLen = len;
        }
        if (root->right && root->right->val == root->val) {
            len = max(len, rightLen + 1);
            currLen += rightLen + 1;
        }
        
        maxLen = max(maxLen, currLen);
        return len;
    }
};

File: 0601-0700/0687-Longest Univalue Path/0687-Longest Univalue Path.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        maxLen = 0
        def dfs(root):
            if root is None:
                return -1
            nonlocal maxLen
            currLen = length = 0
            if root.left:
                leftLen = dfs(root.left)
                if root.left.val == root.val:
                    currLen += 1 + leftLen
                    length = 1 + leftLen
            if root.right:
                rightLen = dfs(root.right)
                if root.right.val == root.val:
                    currLen += 1 + rightLen
                    length = max(length, 1 + rightLen)
            maxLen = max(maxLen, currLen)
            return length
        
        dfs(root)
        return maxLen
